kubernetes/resource_kubernetes_deployment.go

```diff
@@ -0,0 +1,325 @@
+		Read:   resourceKubernetesDeploymentRead,
+		Exists: resourceKubernetesDeploymentExists,
+		Update: resourceKubernetesDeploymentUpdate,
+		Delete: resourceKubernetesDeploymentDelete,
+		Importer: &schema.ResourceImporter{
+			State: schema.ImportStatePassthrough,
+		},
+
+		Timeouts: &schema.ResourceTimeout{
+			Create: schema.DefaultTimeout(10 * time.Minute),
+			Update: schema.DefaultTimeout(10 * time.Minute),
+			Delete: schema.DefaultTimeout(10 * time.Minute),
+		},
+
+		Schema: map[string]*schema.Schema{
+			"metadata": namespacedMetadataSchema("deployment", true),
+			"spec": {
+				Type:        schema.TypeList,
+				Description: "Spec defines the specification of the desired behavior of the deployment. More info: http://releases.k8s.io/HEAD/docs/devel/api-conventions.md#spec-and-status",
+				Required:    true,
+				MaxItems:    1,
+				Elem: &schema.Resource{
+					Schema: map[string]*schema.Schema{
+						"min_ready_seconds": {
+							Type:         schema.TypeInt,
+							Description:  "Minimum number of seconds for which a newly created pod should be ready without any of its container crashing, for it to be considered available. Defaults to 0 (pod will be considered available as soon as it is ready)",
+							Optional:     true,
+							Default:      0,
+							ValidateFunc: validatePositiveInteger,
+						},
+						"paused": {
+							Type:        schema.TypeBool,
+							Description: "Whether the deployment is paused",
+							Optional:    true,
+							Default:     false,
+						},
+						"progress_deadline_seconds": {
+							Type:         schema.TypeInt,
+							Description:  "The maximum time in seconds for a deployment to make progress before it is considered to be failed.",
+							Optional:     true,
+							Default:      600,
+							ValidateFunc: validatePositiveInteger,
+						},
+						"replicas": {
+							Type:         schema.TypeInt,
+							Description:  "The number of desired replicas. Defaults to 1.",
+							Optional:     true,
+							Default:      1,
+							ValidateFunc: validatePositiveInteger,
+						},
+						"revision_history_limit": {
+							Type:         schema.TypeInt,
+							Description:  "The number of old ReplicaSets to retain to allow rollback.",
+							Optional:     true,
+							Default:      10,
+							ValidateFunc: validatePositiveInteger,
+						},
+						"selector": labelSelectorSchema("pods"),
+						"strategy": {
+							Type:        schema.TypeList,
+							Description: "The deployment strategy to use to replace existing pods with new ones.",
+							Optional:    true,
+							MaxItems:    1,
+							Elem: &schema.Resource{
+								Schema: map[string]*schema.Schema{
+									"rolling_update": {
+										Type:        schema.TypeList,
+										Description: "Rolling update config params. Present only if strategy type = 'RollingUpdate'.",
+										Optional:    true,
+										MaxItems:    1,
+										Elem: &schema.Resource{
+											Schema: map[string]*schema.Schema{
```

Comment:
One thing I could use some help on is the best way to handle validation of one schema value in reference to another, like these two parameters here. Valid values for `max_surge` depend on the current value of `max_available`, and vice versa.

Also, these are a weird `intOrString` variant types, I'm not sure the best way to parse, store, or validate these...